ARCHITECTURE.md

```diff
@@ -0,0 +1,122 @@
+
+Ultimately this transformation from a template AST to a TypeScript AST produces a TypeScript module (an extension of the base TypeScript module when there is one) which embeds in the template’s position a set of function calls representing the kind of thing emitted by each operation in the template.
+
+This synthesized module is never persisted to disk.[^debug] Instead, it is presented to the TS compiler, either via a programmatically constructed compiler instance for command line runs or via a language server instance for code editor operations. The results from the TS check are then mapped back to the original template provide diagnostics in the correct location, and sometimes further massaged or rewritten to make them make sense in terms of templates rather than the TS representation of templates.
+
+[^debug]: It is possible to write to disk with a debug flag, but the statement holds for all normal operations.
+
+## Code structure
+
+Glint is composed of the following major components:
+
+
+### Config
+
+**Role:** a library which handles interpreting Glint configuration: which modules should be part of the transform, and how should they be interpreted?
+
+**Invariants:** should know nothing about the rest of the pipeline; it only needs to understand how to parse a Glint configuration and hand it off to the rest of the pipeline.
+
+**Home:** `packages/config`
+
+**Package name:** `@glint/config`
+
+
+### Template DSL
+
+**Role:** a set of TypeScript type definitions which Glint uses as the TypeScript representation of the constituent elements of a template: content, elements, components, expressions, `...attributes`, modifiers, and templates themselves (with or without backing class contexts).
+
+The DSL defines three broad things:
+
+- What it means to *emit* some template content, i.e. what the semantics of emitting `<div {{modifier}}>` are as a kind of TS function application.
+
+- How to *resolve* some template element: when we see `{{this.foo}}`, is that a value to "emit" directly, or is it actually a no-argument function invocation?
+
+- A host of type definitions. Some of these are internal and used to make *emit* and *resolve* work, while others are public types, usable by consumers for authoring their apps. These definitions includes the definitions for types authors can use for writing things like partially applied components.
+
+Note: The fact that the synthesized module is *not* persisted to disk means that some things you might expect to be useful navigation tools within the repo, like the TypeScript *Find All References* command, will not find any references to these.
```

Comment:
Exactly, but/and also for all the `emitContent()` etc. types.

Reply:
Gotcha. On first read I thought this was saying _Find All References_ didn't work on things _used_ in template (like a component) because synthesized modules aren't persisted to disk, rather than the DSL entities.

Reply:
I could not come up with a version of it that was clearer—suggestions hyper-welcome.

Reply:
Maybe just something like this? Ultimately I think it was poor reading comprehension on my part, but I'd lost track of the antecedent of "these" by the time I got to the end of that sentence.

```suggestion
Note: The fact that the synthesized module is *not* persisted to disk means that some things you might expect to be useful navigation tools within the repo, like the TypeScript *Find All References* command, will not find any references to these DSL definitions.
```